docs/lab/02/02.md

```diff
@@ -48,17 +48,187 @@ It is the role of the *IO Bank0* to multiplex and connect the peripherals to the
+the entrypoint and bootloader.
+
+```rust
+use embassy_executor::Spawner;
+
+#[embassy_executor::main]
+async fn main(_spawner: Spawner) {
+    let peripherals = embassy_rp::init(Default::default());
+}
+```
+
+Instead of using the `entry` macro, embassy-rs provides the `embassy_executor::main` macro that
+sets the `main` function as the startup point and defines the correct bootloader.
+
+The `embassy_rp::init` function takes care of the peripheral initialization so that developers can jump
+right in and use them.
+
+:::note
+
+Embassy-rs is designed to work in an asynchronous way and this is why the `main` function is defined as `async`. For the time being, just take it as a must and ignore it.
+
+:::
+
+:::info
+
+The `main` function does not have to loop indefinitely, it is allowed to return, as it runs within the embassy-rs framework.
+
+:::
+
+### Configure GPIO Output
+
+In bare metal and PAC modes, developers have to manually initialize the *IO Bank0* and *SIO* peripherals.
+Embassy-rs provides one single function that returns the GPIO `Output` pin and hides the configuration
+details from developers.
+
+The `pin` variable implements the embadded HAL [`OutputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.OutputPin.html) trait.
+
+```rust
+use gpio::{Level, Output};
+
+// initialize PIN_n (replace n with a number) and set its
+// default value to LOW (0)
+let mut pin = Output::new(peripherals.PIN_n, Level::Low);
+
+// set the pin value to HIGH (1)
+pin.set_high();
+
+// set the pin value to LOW (0)
+pin.set_low();
+```
+
+:::tip
+
+While the device initialization is specific to every hardware device (the example uses the 
+`embassy_rp` crate that is for RP2040), the pin initialization and usage is portable. It
+uses the same code, regardless of the MCU used.
+
+:::
+
+### Configure GPIO Input
+
+Using a pin as input is very similar to using it as output.
+
+```rust
+use gpio::{Input, Pull};
+
+let pin = Input::new(peripherals.PIN_n, Pull::Up);
+
+if pin.is_high() {
+    // Do something if the pin value is HIGH (1)
+} else {
+    // Do something if the pin value if LOW (0)
+}
+```
+
+:::warning 
+
+For a correct use of the buttons, use pull-up, pull-down resistors depending on the mode of operation of the button. Check [01 - Hardware Introduction](https://embedded-rust-101.wyliodrin.com/docs/lab/01#buttons)
+
+:::
+
+### Waiting
+
+Embassy-rs provides support for suspending the execution of the software for an amount of time. It uses
+the [`Timer`](https://docs.rs/embassy-time/0.3.0/embassy_time/struct.Timer.html) structure from the
+[`embassy_time`](https://docs.rs/embassy-time/latest/embassy_time/) crate.
+
+```rust
+// suspend the execution for a period of time
+use embassy_time::Timer;
+
+Timer::after_secs(1).await;
+```
+
+:::tip
+
+If the MCU provides timers, Embassy-rs will use them to suspend the software. This is much more efficient.
```

Comment:
```suggestion
If the MCU provides timers, Embassy-rs will use them to suspend the software. This is very efficient.
```

Reply:
At bate metal explain that looping with `nop` is not efficient.

Reply:
Done